src/v/archival/upload_housekeeping_service.cc

```diff
@@ -0,0 +1,241 @@
+        _workflow.resume(false);
+    }
+}
+
+void upload_housekeeping_service::epoch_timer_callback() {
+    vlog(_ctxlog.info, "Cloud storage housekeeping epoch");
+    if (_workflow.state() != housekeeping_state::draining) {
+        vlog(
+          _ctxlog.info, "Housekeeping epoch timeout, draining the job queue");
+        _workflow.resume(true);
+    }
+}
+
+void upload_housekeeping_service::register_jobs(
+  std::vector<std::reference_wrapper<housekeeping_job>> jobs) {
+    for (auto ref : jobs) {
+        _workflow.register_job(ref.get());
+    }
+}
+
+void upload_housekeeping_service::deregister_jobs(
+  std::vector<std::reference_wrapper<housekeeping_job>> jobs) {
+    for (auto ref : jobs) {
+        _workflow.deregister_job(ref.get());
+    }
+}
+
+housekeeping_workflow::housekeeping_workflow(retry_chain_node& parent)
+  : _parent(parent) {}
+
+void housekeeping_workflow::register_job(housekeeping_job& job) {
+    _pending.push_back(job);
+    _current_backlog++;
+}
+
+void housekeeping_workflow::deregister_job(housekeeping_job& job) {
+    if (_current_job.has_value() && &(_current_job->get()) == &job) {
+        // Current job is running
+        _current_job->get().interrupt();
+    } else {
+        job.interrupt();
+        job._hook.unlink();
+    }
+    _current_backlog--;
+}
+
+void housekeeping_workflow::start() {
+    ssx::spawn_with_gate(_gate, [this] { return run_jobs_bg(); });
+}
+
+ss::future<> housekeeping_workflow::run_jobs_bg() {
+    ss::gate::holder h(_gate);
+    while (!_as.abort_requested()) {
+        vlog(
+          archival_log.debug,
+          "housekeeping_workflow, BG job, state: {}, backlog: {}",
+          _state,
+          _current_backlog);
+        // When the state is active or draining
+        // the loop is processing jobs in round-robin fashion until the
+        // backlog size reaches zero. After that the status changes to idle and
+        // backlog size to the total number of housekeeping jobs.
+        co_await _cvar.wait([this] {
+            return _current_backlog > 0
+                   && (_state == housekeeping_state::active || _state == housekeeping_state::draining);
+        });
+        auto& top = _pending.front();
+        top._hook.unlink();
+        _current_job = std::ref(top);
+        co_await top.run(_parent);
+        _current_job.reset();
+        if (!top.interrupted()) {
+            // If the job was interrupted it's never returned
+            // to the list of pending jobs and never accessd by
+            // the workflow.
+            _pending.push_back(top);
+        }
+        _current_backlog--;
+        if (_current_backlog == 0) {
+            vlog(
+              archival_log.debug,
+              "housekeeping_workflow, transition to idle state from {}",
+              _state);
+            _state = housekeeping_state::idle;
+            _current_backlog = _pending.size();
+        }
+    }
+}
+
+void housekeeping_workflow::resume(bool drain) {
```

Comment:
Explicitly idling the housekeeping flow when there's other work is interesting. I'm curious though -- in other places we use different scheduling groups to indicate what gets priority, don't we? Could we have relied on that to define a  of second-class uploader services, and not have built in this explicit idling?